lib/src/package_analysis/common.dart

```diff
@@ -1,12 +1,204 @@
+  print('path for $dependencyName in $packageLocation is $dependencyFilePath');
+  return dependencyFilePath == null ? null : path.dirname(dependencyFilePath);
+}
+
+/// Recursively fetch all the files in the given folder
+List<resource.File> getAllFiles(resource.Folder folder) {
+  final files = <resource.File>[];
+  for (final child in folder.getChildren()) {
+    if (child is resource.File) {
+      files.add(child);
+    } else if (child is resource.Folder) {
+      files.addAll(getAllFiles(child));
+    } else {
+      throw StateError(
+          'Failed to determine subclass of Resource ${child.path}.');
+    }
+  }
+  return files;
+}
+
+/// Fetch all the hosted dependencies used by a package at a given location
+Future<Map<String, HostedDependency>> getHostedDependencies(
+    String targetPackageDir) async {
+  // fetch a Map of all dependencies from the target package folder
+  final allDependencies = Pubspec.parseYaml(
+          await File(path.join(targetPackageDir, 'pubspec.yaml'))
+              .readAsString())
+      .dependencies;
+
+  // ensure that these dependencies can be found on pub.dev and has version constraints
+  allDependencies.removeWhere((key, value) => value is! HostedDependency);
+  return Map<String, HostedDependency>.from(allDependencies);
+}
+
+/// Given the location of a package and one of its dependencies, return the
+/// dependency version which is installed.
+Future<Version> getInstalledVersion({
+  required PackageAnalysisContext packageAnalysisContext,
+  required String packageLocation,
+  required String dependencyName,
+}) async {
+  // find where this dependency was installed for the target package
+  final dependencyUri = Uri.parse('package:$dependencyName/');
+  final dependencyFilePath = packageAnalysisContext.uriToPath(dependencyUri);
+  // could not resolve uri
+  if (dependencyFilePath == null) {
+    throw Exception(
+        'Could not find package directory of dependency $dependencyName.');
+  }
+
+  // fetch the installed version for this dependency
+  final dependencyPubspecLocation = path.join(
+    path.dirname(dependencyFilePath),
+    'pubspec.yaml',
+  );
+  final dependencyPubspec =
+      Pubspec.parseYaml(await File(dependencyPubspecLocation).readAsString());
+  return dependencyPubspec.version!;
+}
+
+/// Given a version constraint and a list of versions
+/// **from a sorted list of [Version]s in increasing order**, return the the
+/// minimum [Version] allowed by the constraint, or null if no version in the
+/// list is allowed by the constraint.
+Version? findMinAllowedVersion({
+  required VersionConstraint constraint,
+  required List<Version> versions,
+}) {
+  // this first case is common
+  if (constraint is VersionRange &&
+      constraint.includeMin &&
+      versions.contains(constraint.min!)) {
+    return constraint.min!;
+  }
+
+  for (final version in versions) {
+    if (constraint.allows(version)) {
+      return version;
+    }
+  }
+  return null;
```

Comment:
You might not even need a method for doing this :D
```suggestion
  return versions.firstWhere(constraint.allows, orElse: () => null);
```